nntrainer/src/databuffer.cpp

```diff
@@ -106,26 +175,36 @@ void DataBuffer::clear(BufferType type) {
     this->train_data_label.clear();
     this->cur_train_bufsize = 0;
     this->rest_train = max_train;
-    trainReadyFlag = false;
+    trainReadyFlag = DATA_NOT_READY;
     this->train_running = true;
+    if (validation[DATA_TRAIN])
+      if (true == train_thread.joinable())
+        train_thread.join();
   } break;
   case BUF_VAL: {
     val_running = false;
     this->val_data.clear();
     this->val_data_label.clear();
     this->cur_val_bufsize = 0;
     this->rest_val = max_val;
-    valReadyFlag = false;
+    valReadyFlag = DATA_NOT_READY;
     this->val_running = true;
+    if (validation[DATA_TRAIN])
```

Comment:
`validation[DATA_VAL]` looks right (and for line 205). I think this might have caused abort problem. `val_thread` cannot `join` for some cases.

Reply:
theads are only created when these variables true.